File: scripts/create-iframe.py
#!/usr/bin/env python3
"""
Generate iframe HTML for embedding Ozen viewer with audio in Quarto/Jupyter.

Usage:
    python create-iframe.py audio.wav
    python create-iframe.py audio.wav --overlays pitch,formants,hnr
    python create-iframe.py samples/audio.wav --viewer-url ./ozen-web/viewer.html
    python create-iframe.py audio.wav --overlays pitch,formants --height 800
    python create-iframe.py audio.wav --overlays pitch,formants --height 80%

The script calculates the correct relative path from the viewer to the audio file.

IMPORTANT: To view the generated HTML, serve it over HTTP:
    python -m http.server 8000
    # Or use: python scripts/serve-quarto.py

Browsers block file:// iframes for security.
"""

import sys
from pathlib import Path
from urllib.parse import quote

def calculate_relative_path(audio_path, viewer_url):
    """
    Calculate the relative path from the viewer's directory to the audio file.

    Args:
        audio_path: Path to audio file (relative to current directory)
        viewer_url: Path to viewer.html (e.g., "./ozen-web/viewer.html")

    Returns:
        Relative path from viewer's directory to audio file
    """
    # Parse paths
    audio_file = Path(audio_path).resolve()
    viewer_file = Path(viewer_url).resolve()
    viewer_dir = viewer_file.parent

    # Calculate relative path from viewer's directory to audio
    try:
        rel_path = audio_file.relative_to(viewer_dir)
        return str(rel_path)
    except ValueError:
        # Not in same tree, need to go up
        # Calculate relative path using standard algorithm
        audio_parts = audio_file.parts
        viewer_parts = viewer_dir.parts

        # Find common ancestor
        common_length = 0
        for a, b in zip(audio_parts, viewer_parts):
            if a == b:
                common_length += 1
            else:
                break

        # Build path: go up from viewer_dir, then down to audio
        ups = len(viewer_parts) - common_length
        downs = audio_parts[common_length:]

        path_parts = ['..'] * ups + list(downs)
        return '/'.join(path_parts)

def create_embedded_viewer(audio_path, overlays="pitch,formants", viewer_url="./ozen-web/viewer.html", height=600):
    """
    Create iframe HTML with audio path.

    Args:
        audio_path: Path to audio file
        overlays: Comma-separated list of overlays (default: "pitch,formants")
        viewer_url: Path to viewer.html (default: "./ozen-web/viewer.html")
        height: Iframe height - int (pixels) or str (e.g., "80%") (default: 600)

    Returns:
        HTML string for iframe
    """
    audio_file = Path(audio_path)

    if not audio_file.exists():
        raise FileNotFoundError(f"Audio file not found: {audio_path}")

    # Calculate relative path from viewer to audio
    audio_relative = calculate_relative_path(audio_path, viewer_url)

    # URL encode the path (keep slashes for readability)
    encoded_path = quote(audio_relative, safe='/')

    # Format height - convert int to string, keep strings as-is
    height_str = str(height)

    # Create iframe with data-external="1" to prevent Quarto from embedding it as data URL
    iframe = f'''<iframe
  data-external="1"
  src="{viewer_url}?audio={encoded_path}&overlays={overlays}"
  width="100%"
  height="{height_str}"
  frameborder="0"
  style="border: 1px solid #ddd; border-radius: 4px;">
</iframe>'''

    return iframe

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python create-iframe.py <audio-file> [--overlays pitch,formants] [--viewer-url ./ozen-web/viewer.html] [--height 600]")
        print("\nExamples:")
        print("  python create-iframe.py audio.wav --overlays pitch,formants,hnr --height 800")
        print("  python create-iframe.py audio.wav --overlays pitch,formants --height 80%")
        sys.exit(1)

    audio_path = sys.argv[1]
    overlays = "pitch,formants"
    viewer_url = "./ozen-web/viewer.html"
    height = 600

    # Parse optional arguments
    i = 2
    while i < len(sys.argv):
        if sys.argv[i] == "--overlays" and i + 1 < len(sys.argv):
            overlays = sys.argv[i + 1]
            i += 2
        elif sys.argv[i] == "--viewer-url" and i + 1 < len(sys.argv):
            viewer_url = sys.argv[i + 1]
            i += 2
        elif sys.argv[i] == "--height" and i + 1 < len(sys.argv):
            # Height can be numeric (pixels) or string (percentage)
            height_arg = sys.argv[i + 1]
            # Try to convert to int if numeric, otherwise keep as string
            try:
                height = int(height_arg)
            except ValueError:
                height = height_arg
            i += 2
        else:
            print(f"Unknown argument: {sys.argv[i]}", file=sys.stderr)
            sys.exit(1)

    try:
        html = create_embedded_viewer(audio_path, overlays, viewer_url, height)
        print(html)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

File: scripts/create-iframe.R
#!/usr/bin/env Rscript
#' Generate iframe HTML for embedding Ozen viewer with audio in R Markdown/Quarto
#'
#' Command-line usage:
#'   Rscript create-iframe.R audio.wav
#'   Rscript create-iframe.R audio.wav "pitch,formants,hnr"
#'   Rscript create-iframe.R audio.wav "pitch,formants" "./ozen-web/viewer.html"
#'   Rscript create-iframe.R audio.wav "pitch,formants" "./ozen-web/viewer.html" 800
#'   Rscript create-iframe.R audio.wav "pitch,formants" "./ozen-web/viewer.html" "80%"
#'
#' Usage in R Markdown/Quarto:
#'   source("scripts/create-iframe.R")
#'   html <- create_embedded_viewer("audio.wav", overlays = "pitch,formants,hnr")
#'   html <- create_embedded_viewer("audio.wav", overlays = "pitch,formants", height = 800)
#'   html <- create_embedded_viewer("audio.wav", overlays = "pitch,formants", height = "80%")
#'   htmltools::HTML(html)
#'
#' IMPORTANT: To view the generated HTML locally, you must serve it over HTTP:
#'   source("scripts/serve-quarto.R")
#'   serve_quarto()  # Serves on http://localhost:8000
#'
#' Browsers block file:// iframes for security, so double-clicking the HTML won't work.

calculate_relative_path <- function(audio_path, viewer_url) {
  #' Calculate the relative path from the viewer's directory to the audio file.
  #'
  #' @param audio_path Path to audio file (relative to current directory)
  #' @param viewer_url Path to viewer.html (e.g., "./ozen-web/viewer.html")
  #' @return Relative path from viewer's directory to audio file

  # Clean up paths (remove ./ prefix, normalize slashes)
  audio_clean <- gsub("^\\./", "", audio_path)
  viewer_clean <- gsub("^\\./", "", viewer_url)

  # Get viewer directory depth
  viewer_parts <- strsplit(viewer_clean, "/")[[1]]
  viewer_dir_depth <- length(viewer_parts) - 1  # -1 for the filename

  # Build relative path: go up from viewer dir, then to audio
  if (viewer_dir_depth > 0) {
    ups <- paste(rep("..", viewer_dir_depth), collapse = "/")
    relative_path <- paste(ups, audio_clean, sep = "/")
  } else {
    # Viewer is in current directory
    relative_path <- audio_clean
  }

  return(relative_path)
}

create_embedded_viewer <- function(audio_path, overlays = "pitch,formants", viewer_url = "./ozen-web/viewer.html", height = 600) {
  #' Create iframe HTML with audio path
  #'
  #' @param audio_path Path to audio file
  #' @param overlays Comma-separated list of overlays (default: "pitch,formants")
  #' @param viewer_url Path to viewer.html (default: "./ozen-web/viewer.html")
  #' @param height Iframe height: number (pixels) or string (e.g., "80%") (default: 600)
  #' @return HTML string for iframe

  # Check file exists
  if (!file.exists(audio_path)) {
    stop(paste("Audio file not found:", audio_path))
  }

  # Calculate relative path from viewer to audio
  audio_relative <- calculate_relative_path(audio_path, viewer_url)

  # URL encode (keep slashes for readability)
  encoded <- URLencode(audio_relative, reserved = FALSE)

  # Format height - convert number to string, keep strings as-is
  height_str <- as.character(height)

  # Create iframe with data-external="1" to prevent Quarto from embedding it as data URL
  iframe <- sprintf(
    '<iframe
  data-external="1"
  src="%s?audio=%s&overlays=%s"
  width="100%%"
  height="%s"
  frameborder="0"
  style="border: 1px solid #ddd; border-radius: 4px;">
</iframe>',
    viewer_url,
    encoded,
    overlays,
    height_str
  )

  return(iframe)
}

# Command-line interface (only runs when executed as a script, not when sourced)
if (sys.nframe() == 0) {
  args <- commandArgs(trailingOnly = TRUE)

  if (length(args) < 1) {
    cat("Usage: Rscript create-iframe.R <audio-file> [overlays] [viewer-url] [height]\n")
    cat("Example: Rscript create-iframe.R audio.wav 'pitch,formants,hnr' './ozen-web/viewer.html' 800\n")
    cat("         Rscript create-iframe.R audio.wav 'pitch,formants' './ozen-web/viewer.html' '80%'\n")
    quit(status = 1)
  }

  audio_path <- args[1]
  overlays <- if (length(args) >= 2) args[2] else "pitch,formants"
  viewer_url <- if (length(args) >= 3) args[3] else "./ozen-web/viewer.html"
  # Height can be numeric (pixels) or string (percentage)
  height <- if (length(args) >= 4) args[4] else 600

  tryCatch({
    html <- create_embedded_viewer(audio_path, overlays, viewer_url, height)
    cat(html, "\n")
  }, error = function(e) {
    cat("Error:", conditionMessage(e), "\n", file = stderr())
    quit(status = 1)
  })
}
